### PR TITLE
Handle TailCallException in Runtime Instrument

### LINKED ISSUE #427: No Visualization Updates For Function Call In Last Line
### Steps to Reproduce

Consider the following:
```
foo a b = a + b

main = 
    this.foo 1 2
```

Load a project with such Main.enso file and open visualization on `this.foo 1 2` node.

### Expected Result
Visualization update is received by IDE.

### Actual Result
No visualization update comes.

However, if one more line is added, eg
```
foo a b = a + b

main = 
    this.foo 1 2
    5 + 50
```
then it starts working as expected.

### Enso Version
a59f82b821f64ee064a489c9e846b4e7af9f95a0

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #427 

PR implements `TailCallException` handler in the `IdExecutionInstrument` sending correct value updates.

- implement `onReturnExceptional` of the runtime instrument
- add `onExceptionalCallback` to the runtime instrument

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala [hunk 1]
```diff
@@ -457,46 +647,121 @@ class RuntimeServerTest
...
             )
           )
         )
       )
     )
-    context.consumeOut shouldEqual List("1337")
+  }
 
-    // Modify the file
+  it should "send updates from multi liner" in {
```
Comment (kustosz) [inline engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala]: name this better, `multi liner` sounds strangely gangster for a thing existing in a programming language :D

File: engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala [hunk 2]
```diff
@@ -441,6 +529,108 @@ class RuntimeServerTest
...
+  it should "perform operations with State" in {
+    val contextId  = UUID.randomUUID()
+    val requestId  = UUID.randomUUID()
+    val moduleName = "Test.Main"
+
+    val metadata  = new Metadata
+    val idMain    = metadata.addItem(7, 27)
+    val idMainBar = metadata.addItem(26, 8)
+
+    val code =
```
Comment (kustosz) [inline engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala:578]: Please extend the testing scenario, by entering `bar` and verifying that state is indeed correct in a frame detached from the original call. Moreover, pass something non-trivial (I recommend 42) to `State.run` and then start `bar` with a `State.get` and verify that it indeed returned a 42.

**Review by kustosz — APPROVED** [review]
Cool stuff, please test it deeper though to make sure we handle state correctly.
Re: "node errors should be localized". I imagine this can be a larger change, but one that should happen. When something throws an error, the IDE should color a particular node red. Therefore it needs information that the particular node thrown, not just "hey there's an error oopsie". Please at least open an issue for it.

**Comment by kustosz** [comment]
Also, if you decide to implement localized errors as part of this PR, consider the review red, I'd like to see that part :P

**Comment by 4e6** [comment]
Created separate issue #1069 regarding the localization of the interpreter errors.

**Pull request merged**